tree.go

```diff
@@ -5,18 +5,28 @@ import (
 	"errors"
 	"io"
 	"os"
-	"path"
 	"strconv"
 	"strings"
 
 	"gopkg.in/src-d/go-git.v3/core"
 )
 
+const (
+	maxTreeDepth = 1024
+)
+
+// New errors defined by this package.
+var (
+	ErrMaxTreeDepth = errors.New("maximum tree depth exceeded")
+	ErrFileNotFound = errors.New("file not found")
+)
+
 // Tree is basically like a directory - it references a bunch of other trees
 // and/or blobs (i.e. files and sub-directories)
 type Tree struct {
-	Entries map[string]TreeEntry
-	Hash    core.Hash
+	Entries      map[string]TreeEntry
+	OrderedNames []string
```

Comment:
I thought it would be useful and appropriate to keep the entries ordered. I already found it useful for writing tests, and it could make `Tree` comparisons more efficient when we can rely on a known order.

The hash of the tree is already dependent on its order, and that order generally seems well defined. If a tree is recreated with the same entries in a different order I expect that it would produce a different hash. Users of `go-git` might be surprised to get the files back in a weird order when the order is well defined by the tree itself.

Relying on the map here meant that the resulting order would always be random. Users could sort it again to put it back in order, but that seems wasteful. Also, golang's default lexicographic sort algorithm is different than what I'm seeing in most git repositories.

The implementation under review here does add some overhead and I actually don't like it very much. What I'd really like to do is this:

1. Change the definition of `Tree.Entries` to `[]TreeEntry`
2. Add a `Tree.Map` property as `map[string]*TreeEntry`
3. Build `Tree.Map` lazily: only on the first call to `Tree.File()` for example (it basically functions as a cache)
4. Update `TreeWalker` to simply iterate through `Tree.Entries` (it wouldn't touch `Tree.Map`)

Making such a change would have these advantages:

* The implementation would be cleaner than the current code under review
* Users of `Tree` that don't need random lookups would never bother generating a map (a performance improvement)

I was hesitant to do this before because it's more invasive, but I'd be happy to update this PR with the changes described above.